### PR TITLE
[llvm-reduce] Remove DIGlobalVariableExpressions from DICompileUnit's globals

### PR DESCRIPTION
**Pull request by OCHyams**
The 'metadata' delta pass will remove !dbg attachments from globals (which are DIGlobalVariableExpression nodes). The DIGlobalVariableExpressions don't get eliminated from the IR however if they are still referenced by the globals field in DICompileUnit.

Teach the 'di-metadata' pass to try removing global variable operands from metadata tuples as well as DINodes.

### CODE REVIEW

File: llvm/test/tools/llvm-reduce/remove-debug-info-nodes.ll [hunk 1]
```diff
@@ -15,18 +15,15 @@
...
 
 
-
 ; CHECK: define void @test() !dbg [[SUBPROG:![0-9]+]]
 ; CHECK: !llvm.module.flags = !{
 
 ; CHECK: !llvm.dbg.cu = !{[[CU:.+]]}
 
-; CHECK-DAG: [[CU]] = distinct !DICompileUnit(language: DW_LANG_C99,{{.*}}, retainedTypes: [[TYPES:![0-9]+]], globals: [[GLOBALS:![0-9]+]]
+; CHECK-DAG: [[CU]] = distinct !DICompileUnit(language: DW_LANG_C99,{{.*}}, retainedTypes: [[TYPES:![0-9]+]], globals: [[EMPTY:![0-9]+]]
```
Comment (ormris) [inline llvm/test/tools/llvm-reduce/remove-debug-info-nodes.ll:23]: EMPTY is assigned twice

File: llvm/test/tools/llvm-reduce/remove-debug-info-nodes.ll [hunk 2]
```diff
@@ -15,18 +15,15 @@
 ; CHECK-INTERESTINGNESS-DAG: [[SUBPROG]] = distinct !DISubprogram(name: "test",
 
 
-
```
Comment (ormris) [inline llvm/test/tools/llvm-reduce/remove-debug-info-nodes.ll:18]: Whitespace

File: llvm/tools/llvm-reduce/deltas/ReduceDIMetadata.cpp [hunk 3]
```diff
@@ -65,12 +65,14 @@ void identifyUninterestingMDNodes(Oracle &O, MDNodeList &MDs) {
...
     for (size_t I = 0; I < Tup->getNumOperands(); ++I) {
       // Ignore any operands that are not DebugInfo metadata nodes.
-      if (isa_and_nonnull<DINode>(Tup->getOperand(I)))
-        // Don't add uninteresting operands to the tuple.
-        if (!O.shouldKeep())
-          continue;
-
-      TN.push_back(Tup->getOperand(I));
+      if (Metadata *Op = Tup->getOperand(I).get()) {
+        if (isa<DINode>(Op) || isa<DIGlobalVariableExpression>(Op)) {
```
Comment (ormris) [inline llvm/tools/llvm-reduce/deltas/ReduceDIMetadata.cpp:69]: I believe it is possible to find nulls in the MD tree. Given the brute-force approach of this pass, I think `isa_and_nonnull` makes more sense.

Comment (OCHyams) [reply llvm/tools/llvm-reduce/deltas/ReduceDIMetadata.cpp:69]: I've guarded it with `if (Metadata *Op = Tup->getOperand(I).get())` above this line

**Comment by OCHyams** [comment]
Thanks @ormris  - I've addressed your comments

File: llvm/tools/llvm-reduce/deltas/ReduceDIMetadata.cpp (hunk 3 above)
Comment (ormris) [reply llvm/tools/llvm-reduce/deltas/ReduceDIMetadata.cpp:69]: Ah yes. That's fine, then. One nit: The braces on this if statement are not necessary.

**Review by ormris — APPROVED** [review]
After one nit, LGTM

**Pull request merged**